crates/forge/tests/cli/test_cmd.rs

```diff
@@ -3649,3 +3650,182 @@ Encountered a total of 1 failing tests, 0 tests succeeded
+
+Suite result: ok. 1 passed; 0 failed; 0 skipped; [ELAPSED]
+
+Ran 1 test suite [ELAPSED]: 1 tests passed, 0 failed, 0 skipped (1 total tests)
+
+"#]
+    ]);
+});
+
+forgetest!(test_eip712_cheatcode_nested, |prj, cmd| {
+    prj.add_source(
+        "Eip712",
+        r#"
+contract Eip712Structs {
+    struct Transaction {
+        Person from;
+        Person to;
+        Asset tx;
+    }
+    struct Person {
+        address wallet;
+        string name;
+    }
+    struct Asset {
+        address token;
+        uint256 amount;
+    }
+}
+    "#,
+    )
+    .unwrap();
+    prj.insert_ds_test();
+    prj.insert_vm();
+    prj.insert_console();
+
+    prj.add_source("Eip712Cheat.sol", r#"
+import "./test.sol";
+import "./Vm.sol";
+
+string constant CANONICAL = "Transaction(Person from,Person to,Asset tx)Asset(address token,uint256 amount)Person(address wallet,string name)";
+
+contract Eip712Test is DSTest {
+    Vm constant vm = Vm(HEVM_ADDRESS);
+
+    function testEip712HashType_byDefinition() public {
+        bytes32 canonicalHash = keccak256(bytes(CANONICAL));
+
+        // Can figure out the canonical type from a messy string representation of the type,
+        // with an invalid order and extra whitespaces
+        bytes32 fromTypeDef = vm.eip712HashType(
+            "Person(address wallet, string name) Asset(address token, uint256 amount) Transaction(Person from, Person to, Asset tx)"
+        );
+        assertEq(fromTypeDef, canonicalHash);
+    }
+
+    function testEip712HashType_byTypeName() public {
+        bytes32 canonicalHash = keccak256(bytes(CANONICAL));
+
+        // Can figure out the canonical type from the previously generated bindings
+        bytes32 fromTypeName = vm.eip712HashType("Transaction");
+        assertEq(fromTypeName, canonicalHash);
+    }
+
+    function testReverts_Eip712HashType_invalidName() public {
+        // Reverts if the input type is not found in the bindings
+        vm._expectCheatcodeRevert();
+        bytes32 fromTypeName = vm.eip712HashType("InvalidTypeName");
+    }
+}
+"#,
+    )
+    .unwrap();
+
+    // cheatcode by type definition can run without bindings
+    cmd.forge_fuse()
+        .args(["test", "--mc", "Eip712Test", "--match-test", "testEip712HashType_byDefinition"])
+        .assert_success();
+
+    let bindings = prj.root().join("utils").join("JsonBindings.sol");
+    prj.update_config(|config| config.fs_permissions.add(PathPermission::read(&bindings)));
+
+    // cheatcode by type name fails if bindings haven't been generated
+    cmd.forge_fuse()
+        .args([
+            "test",
+            "--mc",
+            "Eip712Test",
+            "--match-test",
+            "testEip712HashType_byTypeName",
+        ])
+        .assert_failure()
+        .stdout_eq(str![[r#"
+...
+Ran 1 test for src/Eip712Cheat.sol:Eip712Test
+[FAIL: vm.eip712HashType: failed to read from [..] No such file or directory (os error 2)] testEip712HashType_byTypeName() ([GAS])
```

Comment:
redact the `No such file or directory (os error 2)]` part too as it's different for win